mempool/cat/pool.go

```diff
@@ -0,0 +1,765 @@
+		txmp.metrics.AlreadySeenTxs.Add(1)
+		// The peer has sent us a transaction that we have already seen
+		return nil, ErrTxInMempool
+	}
+
+	// reserve the key
+	if !txmp.store.reserve(key) {
+		txmp.logger.Debug("mempool already attempting to verify and add transaction", "txKey", fmt.Sprintf("%X", key))
+		txmp.PeerHasTx(txInfo.SenderID, key)
+		return nil, ErrTxInMempool
+	}
+	defer txmp.store.release(key)
+
+	// If a precheck hook is defined, call it before invoking the application.
+	if err := txmp.preCheck(tx); err != nil {
+		txmp.metrics.FailedTxs.Add(1)
+		return nil, mempool.ErrPreCheck{Reason: err}
+	}
+
+	// Early exit if the proxy connection has an error.
+	if err := txmp.proxyAppConn.Error(); err != nil {
+		return nil, err
+	}
+
+	// Invoke an ABCI CheckTx for this transaction.
+	rsp, err := txmp.proxyAppConn.CheckTxSync(abci.RequestCheckTx{Tx: tx})
+	if err != nil {
+		return rsp, err
+	}
+	if rsp.Code != abci.CodeTypeOK {
+		txmp.rejectedTxCache.Push(key)
+		txmp.metrics.FailedTxs.Add(1)
+		return rsp, fmt.Errorf("application rejected transaction with code %d and log %s", rsp.Code, rsp.Log)
+	}
+
+	// Create wrapped tx
+	wtx := newWrappedTx(
+		tx, key, txmp.Height(), rsp.GasWanted, rsp.Priority, rsp.Sender,
+	)
+
+	// Perform the post check
+	err = txmp.postCheck(wtx.tx, rsp)
+	if err != nil {
+		txmp.rejectedTxCache.Push(key)
+		txmp.metrics.FailedTxs.Add(1)
+		return rsp, fmt.Errorf("rejected bad transaction after post check: %w", err)
+	}
+
+	// Now we consider the transaction to be valid. Once a transaction is valid, it
+	// can only become invalid if recheckTx is enabled and RecheckTx returns a non zero code
+	if err := txmp.addNewTransaction(wtx, rsp); err != nil {
+		return nil, err
+	}
+	return rsp, nil
+}
+
+// RemoveTxByKey removes the transaction with the specified key from the
+// mempool. It adds it to the rejectedTxCache so it will not be added again
+func (txmp *TxPool) RemoveTxByKey(txKey types.TxKey) error {
+	txmp.removeTxByKey(txKey)
+	txmp.metrics.EvictedTxs.Add(1)
+	return nil
+}
+
+func (txmp *TxPool) removeTxByKey(txKey types.TxKey) {
+	txmp.rejectedTxCache.Push(txKey)
+	_ = txmp.evictedTxs.Pop(txKey)
+	_ = txmp.store.remove(txKey)
+	txmp.seenByPeersSet.RemoveKey(txKey)
+	txmp.metrics.EvictedTxs.Add(1)
+	txmp.broadcastMtx.Lock()
+	defer txmp.broadcastMtx.Unlock()
+	txmp.txsToBeBroadcast = make(map[types.TxKey]struct{})
+}
+
+// Flush purges the contents of the mempool and the cache, leaving both empty.
+// The current height is not modified by this operation.
+func (txmp *TxPool) Flush() {
+	// Remove all the transactions in the list explicitly, so that the sizes
+	// and indexes get updated properly.
+	size := txmp.Size()
+	txmp.store.reset()
+	txmp.seenByPeersSet.Reset()
+	txmp.evictedTxs.Reset()
+	txmp.rejectedTxCache.Reset()
+	txmp.metrics.EvictedTxs.Add(float64(size))
+	txmp.broadcastMtx.Lock()
+	defer txmp.broadcastMtx.Unlock()
+	txmp.txsToBeBroadcast = make(map[types.TxKey]struct{})
```

Comment:
should also ensure that the channel is empty right?

Reply:
Yeah we can also drain the channel